inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,33 +198,48 @@ def post_audiences(self, course, data, active_tab, msg, error):
             active_tab = "tab_audiences"
 
         try:
-            if "upload_audiences" in data or "audiences" in data:
-                errored_students = []
-                if "upload_audiences" in data:
+            if "audiencefile" in data and 'upload_audiences_creation' in data:
+                # get the Werkzeug datastructures.FileStorage object.
+                # The stream of this object is the stream body of the uploaded file.
+                # Furthermore, FileStorage.stream seems to inherit ʻio.BufferedIOBase`, so this stream should be boiled.
+                reader = csv.reader(io.TextIOWrapper(data["audiencefile"], encoding='utf-8'))
+                csv_data = list(reader)  # read everything else into a list of rows
+                auth_method_ids = self.user_manager.get_auth_methods().keys()
+                for line in csv_data:
+                    if len(line) != 4:
+                        msg["audiences"] = _("File wrongly formatted.")
+                        error["audiences"] = True
+                if "audiences" not in error or not error["audiences"]:
+                    stud_list, aud_li, oth_stu, u_info = self.get_user_lists(course)
                     self.database.audiences.delete_many({"courseid": course.get_id()})
-                    audiences = custom_yaml.load(data["audiencefile"].read())
-                else:
-                    audiences = json.loads(data["audiences"])
-
-                for index, new_audience in enumerate(audiences):
-                    # In case of file upload, no id specified
-                    new_audience['_id'] = new_audience['_id'] if '_id' in new_audience else 'None'
-
-                    # Update the audience
-                    audience, errors = self.update_audience(course, new_audience['_id'], new_audience)
-
-                    # If file upload was done, get the default audience id
-                    audienceid = audience['_id']
-                    errored_students += errors
-
-                if len(errored_students) > 0:
-                    msg["audiences"] = _("Changes couldn't be applied for following students :") + "<ul>"
-                    for student in errored_students:
-                        msg["audiences"] += "<li>" + student + "</li>"
-                    msg["audiences"] += "</ul>"
-                    error["audiences"] = True
-                elif not error:
-                    msg["audiences"] = _("Audience updated.")
+                    for user_id, field, role, description in csv_data:
+                        user_id = user_id.strip()
+                        field = field.strip()
+                        role = role.strip()
+                        description = description.strip()
+                        if field not in ["username", "email", "realname"] + list(auth_method_ids) \
+                                or role not in ["student", "tutors"]:
+                            continue
```

Comment:
Maybe we should warn the user of this. There is much chance they will never know the line wasn't parsed.